src/hotspot/share/opto/predicates.cpp

```diff
@@ -381,6 +382,83 @@ bool TemplateAssertionPredicateExpressionNode::is_template_assertion_predicate(N
 
+InitializedAssertionPredicate::InitializedAssertionPredicate(IfNode* template_assertion_predicate, Node* new_init,
+                                                             Node* new_stride, PhaseIdealLoop* phase)
+    : _template_assertion_predicate(template_assertion_predicate),
+      NOT_PRODUCT(_assertion_predicate_type(template_assertion_predicate->assertion_predicate_type()) COMMA)
+      _new_init(new_init),
+      _new_stride(new_stride),
+      _phase(phase) {}
+
+// Create an Initialized Assertion Predicate at the provided control from the _template_assertion_predicate:
+//
+//          init                 stride
+//            |                    |
+//   OpaqueLoopInitNode  OpaqueLoopStrideNode                      _new_init    _new_stride
+//                  \     /                                               \     /
+//                    ...                                                   ...
+//                     |                                                     |
+//                    Bool                                                new Bool
+//                     |                                                     |
+//                  Opaque4                ======>              control   OpaqueInitializedAssertionPredicate
+//                     |                                              \  /
+//                    If                                             new If
+//                  /    \                                           /    \
+//            success     fail path                         new success   new Halt
+//              proj    (Halt or UCT)                           proj
+//
+IfTrueNode* InitializedAssertionPredicate::create(Node* control) {
+  IdealLoopTree* loop = _phase->get_loop(control);
+  OpaqueInitializedAssertionPredicateNode* new_opaque_bool = create_new_bool(control);
+  IfNode* if_node = create_if_node(control, new_opaque_bool, loop);
+  create_halt_path(if_node, loop);
+  return create_success_path(if_node, loop);
+}
+
+// Create a OpaqueInitializedAssertionPredicate with a new Assertion Predicate Expression.
+OpaqueInitializedAssertionPredicateNode* InitializedAssertionPredicate::create_new_bool(Node* control) {
+  Opaque4Node* template_opaque_bool = _template_assertion_predicate->in(1)->as_Opaque4();
+  TemplateAssertionPredicateExpression template_assertion_predicate_expression(template_opaque_bool);
+  Opaque4Node* new_opaque =
+      template_assertion_predicate_expression.clone_and_replace_init_and_stride(_new_init, _new_stride, control, _phase);
+  OpaqueInitializedAssertionPredicateNode* opaque_bool
+      = new OpaqueInitializedAssertionPredicateNode(new_opaque->in(1)->as_Bool(), _phase->C);
+  _phase->register_new_node(opaque_bool, control);
+  return opaque_bool;
+}
+
+IfNode* InitializedAssertionPredicate::create_if_node(Node* control,
+                                                      OpaqueInitializedAssertionPredicateNode* new_opaque_bool,
```

Comment:
I see why you call it "bool". But it does clash a bit with it actually being a "opaque". Maybe we could call it the "assertion_expression", and the type helps us know what it is?

Reply:
I was not super happy with "bool", either. But I think "Assertion Expression" is a good idea!

So, we have the following naming scheme:
- *Template Assertion Predicate Expression* = `Opaque4` + expression with `OpaqueLoop*Nodes` (starting at `BoolNode`)
- *Assertion Expression* = `OpaqueInitializedAssertionPredicateNode` + expression without `OpaqueLoop*Nodes` (starting at `BoolNode`)

I guess it's easier when we just use "Assertion Expression" instead of "Initialized Assertion Predicate Expression" in the context of this class. "Assertion Expression" already sounds like the real expression. So, I guess that's okay.

I've used the "Assertion Expression" in the entire class now.

What do you think?